### ISSUE TITLE
question: How to do DI in life-cycle hooks?

### ISSUE THREAD

**Issue by SanderElias** [angular/router#186]
This is a 3-fold issue.

The [same plunk](http://plnkr.co/edit/6xwCiZAxUrWuMnWDSOpz?p=preview) again, now the file is components/welcome/welcome.js.

relevant code:
```
    WelcomeController.$inject=['auth']
    function WelcomeController() {
        var vm=this;
        //I don't need it here! it feels awkward to do this here, while I really
        // only need 'auth' in my canActivate
        vm.auth = auth; 
       ...
    }
    
    WelcomeController.prototype.canActivate = function () {
        // how can I inject directly into this function?
        var okprom = this.auth; // fetching this from the controller feels wrong.
        okprom.catch(function () {
              ...
        })
        return this.auth; 
    };    
```

unclear to me:
* inject in controller? but this is canActivate, the controller must be instantiated after this hook?
* I probably need stuff like $http and so on in the canActivate and activate hooks, and perhaps a lot of other services too, that I don't need at all in my controller. For this to work now, I have to inject AND hook them up to the controller. This feels awkward!
* aside from the 2 above point, why do I need to put this on prototype? I would to prefer to declare those directly on the controller, as I do with $inject.

**Comment by timkindberg** [comment]
Can you just move canActivate inside the controller function on this instead of on prototype. I'm not sure I like the prototype approach anyway.

**Comment by btford** [comment]
@Tim – yes

**Comment by SanderElias** [comment]
@timkindberg I know I can move it into the controller. That will work, and take care of the injecttion problem. However, it still feels awkward to put a check that needs to be run INSIDE the same thing. Also then I still need to inject stuff into my controller that I just need for (can)Activate, and this does just feel wrong. The whole crux of DI is that you inject it where you need it.
If there is no (practical) way to solve this, so be it, but if it is an oversight, it needs some attention.

**Comment by btford** [comment]
@SanderElias – here's my proposal:

```js
function WelcomeController() {
  // ...
}

// this function becomes injectable
WelcomeController.prototype.canActivate = function (auth, $instruction) {
  var okprom = this.auth; // fetching this from the controller feels wrong.
  okprom.catch(function () {
    // ...
  });
  return this.auth; 
};
WelcomeController.prototype.canActivate.$inject = ['auth', '$instruction'];
```

I don't think I want to do this though:

```js
// this function becomes injectable
WelcomeController.prototype.canActivate = ['auth', '$instruction', function (auth, $instruction) {
  // ...
}];
```

To me it's weird to have an array on the prototype like that.

By the way, it's entirely possible to achieve this behavior yourself if you include `$inject`.

**Comment by SanderElias** [comment]
@btford That looks quite OK. I indeed dislike the array. (I always did) 

But now you are passing some routeRecognizerResults into the arguments, is that something we really might need? Are you forking this over to the injector?

BTW, I know I can do it myself, it's just a couple of lines of code, but I think it belongs in the routing system!

**Comment by btford** [comment]
> But now you are passing some routeRecognizerResults into the arguments, is that something we really might need? Are you forking this over to the injector?

Maybe. I'm happy to be more conservative and just pass in the `$routeParams`, for instance. That's (probably) the only thing an app developer cares about.

> I think it belongs in the routing system!

I agree, especially since the router is informed about lazy loading, but the general DI system is not.

**Comment by SanderElias** [comment]
@btford I just noticed:

>If there's no $inject property, we'd invoke assuming the fn signature is just function ($instruction) { ... }. Make sense? Or too magical/conditional?

It does make sense to me. However I agree that it is too out of line with the rest of angular. Lets keep this consistent with how the rest works, and just let it be something that is available for injection.

**Comment by btford** [comment]
I just played around a bit with a demo app and I agree. Seems silly to add conditional logic there.

**Comment by SanderElias** [comment]
@btford While I seem to have your attention ;)
what about my 3rth point? what is the reason I need to do `WelcomeController.prototype.canActivate = function () {}` in stead of ` WelcomeController.canActivate = function () {}` 
Controllers are usually just an single instance, so why bother with the prototype?

**Comment by btford** [comment]
Good question.

I also kind of like the idea of having it as a static property. This will make it more natural to statically analyze in Angular 2, which we'll need a good story for anyway to make server-side rendering with non-JS languages easy:

```js
@CanActivate({
  someDeclarativeCriteria: 'someValue'
})
class MyComp {
  // ...
}
```

**Comment by SanderElias** [comment]
>I think I'll make both possible, with the instance method taking precedence over the static methog if present.

Doesn't it make more sense to do it the other way around? 
The prototype might be shared among other controllers, while the static method isn't. Or do I have it backwards?

**Comment by btford** [comment]
@SanderElias – Mind opening a new issue for static lifecycle hooks?

Injectable hooks landed in v0.6.0 btw.

**Comment by SanderElias** [comment]
Sure, no problem: #215
